### ISSUE TITLE
[Benchmark] Kimi-k2 decode pcc drops for batch 128

### ISSUE THREAD

**Issue by gengelageTT** [tenstorrent/tt-xla#5558]
The fix for https://github.com/tenstorrent/tt-xla/issues/4565 has been uplifted into tt-xla and now kimi-k2 no longer hangs for batch 128 however the decode pcc drops from `0.99` for batch 64 to `0.50` for batch 128. This is not an issue for kimi-k2.6 as it sees a decode pcc of `1.00` for batch 128 (expected since the test is only 2 layers).

**Comment by gengelageTT** [comment]
After latest tt-mlir uplift kimi-k2.6 also sees low pcc for batch 128 (pcc of 0). Before the uplift it was passing with good pcc locally for batch 128

**Comment by gengelageTT** [comment]
There are a few bugs with `distributed_rms_norm` lowering the decode PCC of this model (both Kimi-k2 and Kimi-k2.6 variants).

This is only seen at **batch 128**, because batch 128 → **batch 32 per device**, which is what enables the fused `distributed_rms_norm` path (input `<1x1x32x896xbf16>`, second-to-last dim == 32). At other batch sizes the batch-per-device is not a multiple of 32, so the op is decomposed instead and the bugs do not appear.

**Baseline — forcing `distributed_rms_norm` to decompose**, 2 decode layers:

| Variant | Decomposed PCC |
| --- | --- |
| Kimi-k2.6 | `0.999826` |
| Kimi-k2 | `0.999553` |


## The bugs

### 1. Wrong core grid (main issue)

The input `<1x1x32x896xbf16>` splits into **28 tiles** (`32x32`). `DistributedRMSNormWidthShardInputRewritePattern`
in tt-mlir places these 28 tiles row-major on an `8×4` core grid (**32 cores**), leaving **4 empty cores** the kernel still reads → garbage output. The tt-metal unit test (`test_distributed_rms_norm_decode_configs.py`) for this op and input shape uses a `4×7` grid (**28 cores**, solid rectangle).

Fixing the core grid to a solid `4×7` rectangle almost restores PCC, but it is still lower than the decomposed version:

| Variant | Fixed grid | Decomposed |
| --- | --- | --- |
| Kimi-k2.6 | `0.991325` | `0.999826` |
| Kimi-k2 | `0.980880` | `0.999553` |

### 2. Kernel compute config doesn't match metal

tt-mlir sets `mathApproxMode=false`, `fp32DestAccEn=true`, `packerL1Acc=true` in the kernel config, however setting the **opposite** (matching the tt-metal unit test's config) actually produces **better** PCC.

Single-op testing determined `mathApproxMode=false` is the main issue but `fp32DestAccEn=true` also contributes. Matching the config metal uses in tt-mlir and fixing bug (1) results in:

| Variant | Metal config | Decomposed |
| --- | --- | --- |
| Kimi-k2.6 | `0.996419` | `0.999826` |
| Kimi-k2 | `0.989951` | `0.999553` |

### 3. The fused kernel has worse PCC than the decomposed path

After fixing bugs (1) and (2) the PCC is still lower when `distributed_rms_norm` is used. Both fused and decomposed degrade as layers are added so the ~`0.01` fused-vs-decomposed gap doesn't clearly accumulate with depth:

| Kimi-k2 | Fused  | Decomposed  |
| --- | --- | --- |
| 2 layers | `0.989951` | `0.999553` |
| 4 layers | `0.988114` | `0.995405` |

Fused distributed RMSNorm (ttnn.fused_rms_minimal / rms_allgather) is measurably less accurate per-op than the decomposed primitives (rms_norm_pre_all_gather + all_gather_async + rms_norm_post_all_gather) on identical inputs against the same fp32 golden.

From **_tests/nightly/tg/ccl/test_rms_fused_vs_decompose_pcc.py_** in tt-metal `gengelage/repro-fused-rmsnorm-pcc`:
|  | Mean PCC (over 10 iterations)|
| --- | --- |
| Decomposed | `0.999997` | 
| Fused | `0.999877` |

**Comment by gengelageTT** [comment]
## Update

### 3. The fused kernel has worse PCC than the decomposed path

This is a bug we can fix in tt-mlir, but it also exists in the tt-metal unit tests. Two different issues: a stats tensor sizing bug and a hardcoded rsqrt mode bug.

#### Metal testing

**Stats tensor sizing.** [tests/nightly/tg/ccl/test_distributed_rms_norm_decode_configs.py:106](https://github.com/tenstorrent/tt-metal/blob/main/tests/nightly/tg/ccl/test_distributed_rms_norm_decode_configs.py#L106) says:
```
# Stats buffer configuration - single core sharding so stats.padded_shape[-1] = num_devices * TILE_WIDTH
# This is critical for correct RMS scaling across devices
```
But the unit test in the same file doesn't follow this, and instead sets:
```
# Create persistent stats tensor
ag_shape = [1, 1, 32, num_devices]
stats_tensor = torch.zeros(ag_shape, dtype=torch.bfloat16)
```
The result is no cross-device averaging (each device normalizes by a single (wrong) device's `E[x²]`).

Correcting this raises PCC from `~0.99988` to `~0.99992`. If we also fix the rsqrt mode (see below), PCC reaches` 0.999998` which matches the decomposed path (yay).

**Hardcoded rsqrt mode.** This is a separate metal bug with approx rsqrt: the decomposed version of the op doesn't see this since it always uses the accurate default (`rsqrt_tile<false>`) - even when `mathApproxMode=true`. If we adjust the fused kernel to do the same then it has good PCC for `mathApproxMode=true` too.

#### MLIR / XLA testing

tt-mlir makes the same **stats-tensor sizing** mistake, hardcoding:
```
SmallVector<int64_t> statsShape = {1, 1, 32, 32};
```
when it should be:
```
SmallVector<int64_t> statsShape = {1, 1, 32, 32 * numDevices};
```
Unfortunately, applying this fix to the full model triggers an L1 OOM that still needs investigation. 


### 2. Kernel compute config doesn't match metal
The compute config used in tt-mlir is actually more accurate once the fix is applied. The stats buffer bug was just amplified at a higher accuracy.

**Comment by jameszianxuTT** [comment]
Also, this op is has broken impl for multihost and cannot be executed under multihost at all.

**Comment by gengelageTT** [comment]
## Update

### 4. (NEW) L1 OOM when running full model with proper stats-tensor size
Increasing the stats-tensor size by num_devices (8 TP in kimi-k2) causes an L1 OOM in `paged_update_cache`. 

With the original (incorrect) stats tensor, the L1 memory before `distributed_rms_norm` increases from `0.005 MB` to `0.009 MB` then it stays at least that high for the rest of execution (until OOM):
```
Device memory state before operation EmptyOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.005 MB, ...}
Device memory state before operation ToLayoutOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.009 MB, ...}
...
Device memory state before operation DistributedRMSNormOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.011 MB, ...}
Device memory state before operation DeallocateOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.013 MB, ...}
Device memory state before operation DeallocateOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.011 MB, ...}
Tensor is retained thus not deallocating. To deallocate, set retain to false first
...
Device memory state before operation ReshapeOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.009 MB, ...}
...
```
When the stats tensor is 8x the size of above (the correct size), the L1 memory before `distributed_rms_norm` increases from `0.005 MB` to `0.037 MB` then it stays at least that high for the rest of execution (until OOM):
```
Device memory state before operation EmptyOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.005 MB, ...}
Device memory state before operation ToLayoutOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.037 MB, ...}
...
Device memory state before operation DistributedRMSNormOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.039 MB, ...}
Device memory state before operation DeallocateOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.041 MB, ...}
Device memory state before operation DeallocateOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.039 MB, ...}
Tensor is retained thus not deallocating. To deallocate, set retain to false first
...
Device memory state before operation ReshapeOp
Device L1 memory state: MemoryView{... totalBytesAllocatedPerBank: 0.037 MB, ...}
...
```

Since the increase in L1 used is proportionate to the increase in stats-tensor size (`0.004 MB x8 = 0.032 MB`) the stats tensor is likely not getting deallocated from L1 which is causing OOM for larger tensor sizes.

**Comment by gengelageTT** [comment]
With fixes for bugs (1) and (3), kimi-k2 (2 layers) passes with good PCC for batch 128 when program cache is disabled:
```
Prefill PCC verification passed with PCC=0.999534, rel_l2=3.142280e-02
First decode PCC verification passed with PCC=0.999550, rel_l2=2.975788e-02
PASSED
```

Which nearly matches the decode PCC when decomposed: `0.999553`

**Comment by gengelageTT** [comment]
Test passes with program cache enabled with changes in tt-metal (`gengelage/test-rms-norm-mlir`) and tt-mlir (`gengelage/kimi-k2-fix-testing-v3`) that allow the stats tensor to be deallocated from L1. Need fix in tt-metal before that stats tensor can be fixed in tt-mlir.
